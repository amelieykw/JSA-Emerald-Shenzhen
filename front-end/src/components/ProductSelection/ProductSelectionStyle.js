import { StyleSheet } from 'react-native'

const common = {
  context: {
    display: 'flex',
    flexDirection: 'row',
    width: '75%'
  },
  title: {
    fontWeight: 'bold'
  },
  icon: {
    fontSize: 40,
    paddingLeft: '10%',
    paddingRight: '10%'
  },
  card: {
    padding: '5%',
    paddingLeft: 0,
    borderBottomWidth: 2,
    borderBottomColor: '#D4D4D4'
  },
  select: {
    marginTop: 0
  },
  button: {
    alignSelf: 'flex-end'
  },
  buttonView: {
    marginTop: '5%'
  }
}

const tinyStyle = StyleSheet.create({
  ...common,
  show: {
    display: 'none'
  }
})

const extendStyle = StyleSheet.create({
  ...common,
  show: {
    display: 'flex'
  }
})

export { tinyStyle, extendStyle }
