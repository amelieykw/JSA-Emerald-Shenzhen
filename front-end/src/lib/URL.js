import { BASE_URL } from 'react-native-dotenv'

const BASE = BASE_URL || 'http://10.22.18.66:3000'

const URL = {
  getLoans: `${BASE}/loans`,
  getDetail: id => `${BASE}/loans/${id}`,
  getProducts: `${BASE}/products`
}

export default URL
