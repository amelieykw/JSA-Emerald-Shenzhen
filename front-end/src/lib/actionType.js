const ACTION_TYPE = {
  INIT_LOANLIST: 'INIT_LOANLIST',
  INIT_LOANLIST_SUCCESS: 'INIT_LOANLIST_SUCCESS',
  INIT_PRODUCTLIST: 'INIT_PRODUCTLIST',
  INIT_PRODUCTLIST_SUCCESS: 'INIT_PRODUCTLIST_SUCCESS'
}

export default ACTION_TYPE
