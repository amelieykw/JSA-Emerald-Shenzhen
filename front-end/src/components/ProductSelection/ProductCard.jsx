import React, { useContext } from 'react'
import PropTypes from 'prop-types'
import { Text, Icon, Button, View } from 'native-base'
import { NavigationContext } from 'react-navigation'
import { tinyStyle, extendStyle } from './ProductSelectionStyle'

export default function ProductCard({ product, isExtend }) {
  const navigation = useContext(NavigationContext)

  function handleSelectClick() {
    navigation.navigate('NewLoan', { product })
  }

  const { name, description } = product
  const style = isExtend ? extendStyle : tinyStyle

  return (
    <View style={style.card} opacity={isExtend ? 1 : 0.7}>
      <View style={style.context}>
        <Icon style={style.icon} name="person" />
        <View>
          <Text style={style.title}>{name}</Text>
          <Text>
            {!isExtend && description.length > 23
              ? `${description.substring(0, 32)}...`
              : description}
          </Text>
        </View>
      </View>
      <View style={{ ...style.show, ...style.buttonView }}>
        <Button
          style={style.button}
          onPress={() => {
            handleSelectClick()
          }}
        >
          <Text>SELECT</Text>
        </Button>
      </View>
    </View>
  )
}

ProductCard.propTypes = {
  product: PropTypes.objectOf(PropTypes.any).isRequired,
  isExtend: PropTypes.bool.isRequired
}
