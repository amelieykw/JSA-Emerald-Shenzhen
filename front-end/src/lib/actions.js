import URL from './URL'

const ACTION_TYPE = {
  LOADING: 'LOADING',
  LOADDONE: 'LOADDONE',
  INIT_LOANLIST: 'INIT_LOANLIST',
  INIT_LOANLIST_SUCCESS: 'INIT_LOANLIST_SUCCESS',
  INIT_PRODUCTLIST: 'INIT_PRODUCTLIST',
  INIT_PRODUCTLIST_SUCCESS: 'INIT_PRODUCTLIST_SUCCESS'
}

async function fetchData(url) {
  console.log(URL)
  console.log(`Outside fetchData return : ${Object.entries(URL)}`)
  console.log(`Outside fetchData return : ${URL.getProducts}`)
  return fetch(url, {
    headers: { Authentication: 'Bearer token', Accept: 'application/json' }
  }).then(res => res.json())
}

function fetchLoanList() {
  return async dispatch => {
    dispatch({ type: ACTION_TYPE.LOADING })
    dispatch({ type: ACTION_TYPE.INIT_LOANLIST })

    try {
      const data = await fetchData(URL.getLoans)
      dispatch({ type: ACTION_TYPE.INIT_LOANLIST_SUCCESS, loanList: data })
      dispatch({ type: ACTION_TYPE.LOADDONE })
    } catch (error) {
      // dispatch({ type: ACTION_TYPE.ERROR, error })
      console.log(error)
    }
  }
}

function fetchProductList() {
  console.log(`Outside fetchProductList try : ${Object.entries(URL)}`)
  console.log(`Outside fetchProductList try : ${URL.getProducts}`)
  return async dispatch => {
    dispatch({ type: ACTION_TYPE.LOADING })
    dispatch({ type: ACTION_TYPE.INIT_PRODUCTLIST })

    try {
      console.log(`Inside fetchProductList try : ${Object.entries(URL)}`)
      console.log(`Inside fetchProductList try : ${URL.getProducts}`)
      const data = await fetchData(URL.getProducts)
      dispatch({ type: ACTION_TYPE.INIT_PRODUCTLIST_SUCCESS, productList: data })
      dispatch({ type: ACTION_TYPE.LOADDONE })
    } catch (error) {
      // dispatch({ type: ACTION_TYPE.ERROR, error })
      console.log(error)
    }
  }
}

export { fetchLoanList, fetchProductList, ACTION_TYPE }
