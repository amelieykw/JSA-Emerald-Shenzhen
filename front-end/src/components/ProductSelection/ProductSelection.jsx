import React, { useState } from 'react'
import { Container, Content, View } from 'native-base'
import { TouchableWithoutFeedback, ScrollView } from 'react-native'
import PropTypes from 'prop-types'

import data from '../../../helpers/fakeProductsData'

import ProductCard from './ProductCard'

export default function ProductSelection({ products }) {
  const [selectId, setSelectId] = useState(-1)

  return (
    <Container>
      <Content>
        <ScrollView>
          {products.map((product, sid) => {
            return (
              <TouchableWithoutFeedback
                key={product.id}
                onPress={() => {
                  setSelectId(sid)
                }}
              >
                <View>
                  <ProductCard isExtend={sid === selectId} product={product} />
                </View>
              </TouchableWithoutFeedback>
            )
          })}
        </ScrollView>
      </Content>
    </Container>
  )
}

ProductSelection.propTypes = {
  products: PropTypes.arrayOf(PropTypes.any)
}

ProductSelection.defaultProps = {
  products: data.products
}

ProductSelection.navigationOptions = {
  title: 'products'
}
