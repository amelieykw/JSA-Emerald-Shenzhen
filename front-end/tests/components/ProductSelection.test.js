import React from 'react'
import configureStore from 'redux-mock-store'
import renderer from 'react-test-renderer'
import { Provider } from 'react-redux'
import thunk from 'redux-thunk'
import ProductSelection from '../../src/components/ProductSelection/ProductSelection'
import data from '../../helpers/mockData_FE'
import { fetchProductList, ACTION_TYPE } from '../../src/lib/actions'

const mockStore = configureStore([thunk])

describe('<ProductSelection />', () => {
  let store
  let component
  let payloads

  beforeEach(() => {
    store = mockStore({ productList: data.products, loading: true })
    payloads = [
      {
        type: ACTION_TYPE.LOADING
      },
      {
        type: ACTION_TYPE.INIT_PRODUCTLIST
      },
      {
        type: ACTION_TYPE.INIT_PRODUCTLIST_SUCCESS,
        productList: {}
      },
      {
        type: ACTION_TYPE.LOADDONE
      }
    ]
  })
  it('<ProductSelection /> renders correctly', () => {
    component = renderer.create(
      <Provider store={store}>
        <ProductSelection />
      </Provider>
    )
    expect(component.toJSON()).toMatchSnapshot()
  })
  it('<ProductSelection > has 1 child', () => {
    expect(component.toJSON().children.length).toBe(1)
  })
  it('<ProductSelection /> should dispatch an action when mounted', async () => {
    renderer.act(() => {
      component = renderer.create(
        <Provider store={store}>
          <ProductSelection />
        </Provider>
      )
    })

    // eslint-disable-next-line no-undef
    window.fetch = jest.fn().mockImplementation(() =>
      Promise.resolve({
        json: () => Promise.resolve({})
      })
    )

    await store.dispatch(fetchProductList())
    const actions = store.getActions()

    expect(actions.pop()).toEqual(payloads.pop())
    expect(actions.pop()).toEqual(payloads.pop())
    expect(actions.pop()).toEqual(payloads.pop())
    expect(actions.pop()).toEqual(payloads.pop())
  })
})
