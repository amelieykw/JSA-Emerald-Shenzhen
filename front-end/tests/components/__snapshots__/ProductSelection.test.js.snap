// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProductSelection /> <ProductSelection /> renders correctly 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#fff",
      "flex": 1,
      "height": 1334,
    }
  }
>
  <RCTScrollView
    automaticallyAdjustContentInsets={false}
    contentContainerStyle={
      Array [
        Object {
          "padding": undefined,
        },
        undefined,
      ]
    }
    contentInset={
      Object {
        "bottom": 0,
      }
    }
    enableAutomaticScroll={true}
    enableOnAndroid={false}
    enableResetScrollToCoords={true}
    extraHeight={75}
    extraScrollHeight={0}
    getScrollResponder={[Function]}
    handleOnScroll={[Function]}
    keyboardDismissMode="interactive"
    keyboardOpeningTime={250}
    keyboardShouldPersistTaps="handled"
    keyboardSpace={0}
    onScroll={[Function]}
    resetKeyboardSpace={[Function]}
    resetScrollToCoords={
      Object {
        "x": 0,
        "y": 0,
      }
    }
    scrollEventThrottle={1}
    scrollForExtraHeightOnAndroid={[Function]}
    scrollIntoView={[Function]}
    scrollToEnd={[Function]}
    scrollToFocusedInput={[Function]}
    scrollToPosition={[Function]}
    showsVerticalScrollIndicator={true}
    style={
      Object {
        "backgroundColor": "transparent",
        "flex": 1,
      }
    }
    update={[Function]}
    viewIsInsideTabBar={false}
  >
    <View>
      <RCTScrollView>
        <View>
          <View
            accessible={true}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={Object {}}
          >
            <View
              opacity={0.7}
              style={
                Array [
                  Object {},
                  Object {
                    "borderBottomColor": "#D4D4D4",
                    "borderBottomWidth": 2,
                    "padding": "5%",
                    "paddingLeft": 0,
                  },
                ]
              }
            >
              <View
                style={
                  Array [
                    Object {},
                    Object {
                      "display": "flex",
                      "flexDirection": "row",
                      "width": "75%",
                    },
                  ]
                }
              >
                <Text />
                <View
                  style={Object {}}
                >
                  <Text
                    style={
                      Array [
                        Object {
                          "color": "#000",
                          "fontFamily": "System",
                          "fontSize": 16,
                        },
                        Object {
                          "fontWeight": "bold",
                        },
                      ]
                    }
                    uppercase={false}
                  >
                    Student Loan
                  </Text>
                  <Text
                    style={
                      Object {
                        "color": "#000",
                        "fontFamily": "System",
                        "fontSize": 16,
                      }
                    }
                    uppercase={false}
                  >
                    Fixed low interests for students...
                  </Text>
                </View>
              </View>
              <View
                style={
                  Array [
                    Object {},
                    Object {
                      "display": "none",
                      "marginTop": "5%",
                    },
                  ]
                }
              >
                <View
                  accessible={true}
                  isTVSelectable={true}
                  onResponderGrant={[Function]}
                  onResponderMove={[Function]}
                  onResponderRelease={[Function]}
                  onResponderTerminate={[Function]}
                  onResponderTerminationRequest={[Function]}
                  onStartShouldSetResponder={[Function]}
                  style={
                    Object {
                      "alignItems": "center",
                      "alignSelf": "flex-end",
                      "backgroundColor": "#007aff",
                      "borderBottomWidth": null,
                      "borderColor": "#007aff",
                      "borderLeftWidth": null,
                      "borderRadius": 5,
                      "borderRightWidth": null,
                      "borderTopWidth": null,
                      "borderWidth": undefined,
                      "elevation": 2,
                      "flexDirection": "row",
                      "height": 45,
                      "justifyContent": "space-between",
                      "opacity": 1,
                      "paddingBottom": 6,
                      "paddingTop": 6,
                      "shadowColor": undefined,
                      "shadowOffset": undefined,
                      "shadowOpacity": undefined,
                      "shadowRadius": undefined,
                    }
                  }
                >
                  <Text
                    style={
                      Object {
                        "backgroundColor": "transparent",
                        "color": "#fff",
                        "fontFamily": "System",
                        "fontSize": 16.5,
                        "marginLeft": 0,
                        "marginRight": 0,
                        "paddingLeft": 16,
                        "paddingRight": 16,
                      }
                    }
                    uppercase={false}
                  >
                    SELECT
                  </Text>
                </View>
              </View>
            </View>
          </View>
          <View
            accessible={true}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={Object {}}
          >
            <View
              opacity={0.7}
              style={
                Array [
                  Object {},
                  Object {
                    "borderBottomColor": "#D4D4D4",
                    "borderBottomWidth": 2,
                    "padding": "5%",
                    "paddingLeft": 0,
                  },
                ]
              }
            >
              <View
                style={
                  Array [
                    Object {},
                    Object {
                      "display": "flex",
                      "flexDirection": "row",
                      "width": "75%",
                    },
                  ]
                }
              >
                <Text />
                <View
                  style={Object {}}
                >
                  <Text
                    style={
                      Array [
                        Object {
                          "color": "#000",
                          "fontFamily": "System",
                          "fontSize": 16,
                        },
                        Object {
                          "fontWeight": "bold",
                        },
                      ]
                    }
                    uppercase={false}
                  >
                    Unsecured
                  </Text>
                  <Text
                    style={
                      Object {
                        "color": "#000",
                        "fontFamily": "System",
                        "fontSize": 16,
                      }
                    }
                    uppercase={false}
                  >
                    Fixed low interests for students...
                  </Text>
                </View>
              </View>
              <View
                style={
                  Array [
                    Object {},
                    Object {
                      "display": "none",
                      "marginTop": "5%",
                    },
                  ]
                }
              >
                <View
                  accessible={true}
                  isTVSelectable={true}
                  onResponderGrant={[Function]}
                  onResponderMove={[Function]}
                  onResponderRelease={[Function]}
                  onResponderTerminate={[Function]}
                  onResponderTerminationRequest={[Function]}
                  onStartShouldSetResponder={[Function]}
                  style={
                    Object {
                      "alignItems": "center",
                      "alignSelf": "flex-end",
                      "backgroundColor": "#007aff",
                      "borderBottomWidth": null,
                      "borderColor": "#007aff",
                      "borderLeftWidth": null,
                      "borderRadius": 5,
                      "borderRightWidth": null,
                      "borderTopWidth": null,
                      "borderWidth": undefined,
                      "elevation": 2,
                      "flexDirection": "row",
                      "height": 45,
                      "justifyContent": "space-between",
                      "opacity": 1,
                      "paddingBottom": 6,
                      "paddingTop": 6,
                      "shadowColor": undefined,
                      "shadowOffset": undefined,
                      "shadowOpacity": undefined,
                      "shadowRadius": undefined,
                    }
                  }
                >
                  <Text
                    style={
                      Object {
                        "backgroundColor": "transparent",
                        "color": "#fff",
                        "fontFamily": "System",
                        "fontSize": 16.5,
                        "marginLeft": 0,
                        "marginRight": 0,
                        "paddingLeft": 16,
                        "paddingRight": 16,
                      }
                    }
                    uppercase={false}
                  >
                    SELECT
                  </Text>
                </View>
              </View>
            </View>
          </View>
          <View
            accessible={true}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={Object {}}
          >
            <View
              opacity={0.7}
              style={
                Array [
                  Object {},
                  Object {
                    "borderBottomColor": "#D4D4D4",
                    "borderBottomWidth": 2,
                    "padding": "5%",
                    "paddingLeft": 0,
                  },
                ]
              }
            >
              <View
                style={
                  Array [
                    Object {},
                    Object {
                      "display": "flex",
                      "flexDirection": "row",
                      "width": "75%",
                    },
                  ]
                }
              >
                <Text />
                <View
                  style={Object {}}
                >
                  <Text
                    style={
                      Array [
                        Object {
                          "color": "#000",
                          "fontFamily": "System",
                          "fontSize": 16,
                        },
                        Object {
                          "fontWeight": "bold",
                        },
                      ]
                    }
                    uppercase={false}
                  >
                    Pay day
                  </Text>
                  <Text
                    style={
                      Object {
                        "color": "#000",
                        "fontFamily": "System",
                        "fontSize": 16,
                      }
                    }
                    uppercase={false}
                  >
                    Fixed low interests for students...
                  </Text>
                </View>
              </View>
              <View
                style={
                  Array [
                    Object {},
                    Object {
                      "display": "none",
                      "marginTop": "5%",
                    },
                  ]
                }
              >
                <View
                  accessible={true}
                  isTVSelectable={true}
                  onResponderGrant={[Function]}
                  onResponderMove={[Function]}
                  onResponderRelease={[Function]}
                  onResponderTerminate={[Function]}
                  onResponderTerminationRequest={[Function]}
                  onStartShouldSetResponder={[Function]}
                  style={
                    Object {
                      "alignItems": "center",
                      "alignSelf": "flex-end",
                      "backgroundColor": "#007aff",
                      "borderBottomWidth": null,
                      "borderColor": "#007aff",
                      "borderLeftWidth": null,
                      "borderRadius": 5,
                      "borderRightWidth": null,
                      "borderTopWidth": null,
                      "borderWidth": undefined,
                      "elevation": 2,
                      "flexDirection": "row",
                      "height": 45,
                      "justifyContent": "space-between",
                      "opacity": 1,
                      "paddingBottom": 6,
                      "paddingTop": 6,
                      "shadowColor": undefined,
                      "shadowOffset": undefined,
                      "shadowOpacity": undefined,
                      "shadowRadius": undefined,
                    }
                  }
                >
                  <Text
                    style={
                      Object {
                        "backgroundColor": "transparent",
                        "color": "#fff",
                        "fontFamily": "System",
                        "fontSize": 16.5,
                        "marginLeft": 0,
                        "marginRight": 0,
                        "paddingLeft": 16,
                        "paddingRight": 16,
                      }
                    }
                    uppercase={false}
                  >
                    SELECT
                  </Text>
                </View>
              </View>
            </View>
          </View>
          <View
            accessible={true}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={Object {}}
          >
            <View
              opacity={0.7}
              style={
                Array [
                  Object {},
                  Object {
                    "borderBottomColor": "#D4D4D4",
                    "borderBottomWidth": 2,
                    "padding": "5%",
                    "paddingLeft": 0,
                  },
                ]
              }
            >
              <View
                style={
                  Array [
                    Object {},
                    Object {
                      "display": "flex",
                      "flexDirection": "row",
                      "width": "75%",
                    },
                  ]
                }
              >
                <Text />
                <View
                  style={Object {}}
                >
                  <Text
                    style={
                      Array [
                        Object {
                          "color": "#000",
                          "fontFamily": "System",
                          "fontSize": 16,
                        },
                        Object {
                          "fontWeight": "bold",
                        },
                      ]
                    }
                    uppercase={false}
                  >
                    Secured
                  </Text>
                  <Text
                    style={
                      Object {
                        "color": "#000",
                        "fontFamily": "System",
                        "fontSize": 16,
                      }
                    }
                    uppercase={false}
                  >
                    Fixed low interests for students...
                  </Text>
                </View>
              </View>
              <View
                style={
                  Array [
                    Object {},
                    Object {
                      "display": "none",
                      "marginTop": "5%",
                    },
                  ]
                }
              >
                <View
                  accessible={true}
                  isTVSelectable={true}
                  onResponderGrant={[Function]}
                  onResponderMove={[Function]}
                  onResponderRelease={[Function]}
                  onResponderTerminate={[Function]}
                  onResponderTerminationRequest={[Function]}
                  onStartShouldSetResponder={[Function]}
                  style={
                    Object {
                      "alignItems": "center",
                      "alignSelf": "flex-end",
                      "backgroundColor": "#007aff",
                      "borderBottomWidth": null,
                      "borderColor": "#007aff",
                      "borderLeftWidth": null,
                      "borderRadius": 5,
                      "borderRightWidth": null,
                      "borderTopWidth": null,
                      "borderWidth": undefined,
                      "elevation": 2,
                      "flexDirection": "row",
                      "height": 45,
                      "justifyContent": "space-between",
                      "opacity": 1,
                      "paddingBottom": 6,
                      "paddingTop": 6,
                      "shadowColor": undefined,
                      "shadowOffset": undefined,
                      "shadowOpacity": undefined,
                      "shadowRadius": undefined,
                    }
                  }
                >
                  <Text
                    style={
                      Object {
                        "backgroundColor": "transparent",
                        "color": "#fff",
                        "fontFamily": "System",
                        "fontSize": 16.5,
                        "marginLeft": 0,
                        "marginRight": 0,
                        "paddingLeft": 16,
                        "paddingRight": 16,
                      }
                    }
                    uppercase={false}
                  >
                    SELECT
                  </Text>
                </View>
              </View>
            </View>
          </View>
        </View>
      </RCTScrollView>
    </View>
  </RCTScrollView>
</View>
`;
